### PR TITLE
[BUG] Removed memory leak to do with games/sessions not being removed from map

### LINKED ISSUE #275: [BUG] Completed Sessions Never Removed
***
#### Describe the bug:
When a session is completed ie a game of yumble is completed, it is possible to still join the room. This is because the game is never removed from the activeGames map in `Games.js`

This is a potential memory leak that could cause the server to crash if not dealt with.

### To Reproduce the bug:
1. Finish a game of yumble
2. Create a new tab and rejoin the same game with the same code
The expected behaviour is that the lobby code should be invalid OR should take you to the results screen of the completed game

Screenshots - 
image
Here you can see a room that has been joined without the host, as the room has been joined after the game has been completed.

### PR DESCRIPTION
**Pull request by UltimaKuma**
Closes #275 
Removes the game from the activeGames map when the game is finished.
Required the creation of an extra socket event as during the preferences stage, there was no reference to which game belonged to which socket, thus making it impossible to reference to delete.
The new socket event `"host_room"` sets the selected room code host to be the socket  caller.

## Before:
Screenshot 2021-04-08 172945 (2)
## After:
Screenshot 2021-04-08 174316
As seen in the heap comparison snapshots above, the change removes the residual `Game`s and `SocketSession`s in memory. 

From my testing, there seem to be no other memory leaks from the server.

## Note
I have also added a new npm script to run the server in inspect mode.

Use `npm run inspect` to run the script.

Assuming that chrome or a chromium based browser is being used, DevTools can be accessed through `chrome://inspect`, which can be useful for accessing memory information from a server being run locally.

### CODE REVIEW

**Review by parfei — COMMENTED** [review]
Looks great, thanks Leon :) Just double checking, do we handle when suddenly someone disconnects/host disconnects/no one is left in the room, and we need to remove games/sessions?

**Comment by UltimaKuma** [comment]
> Looks great, thanks Leon :) Just double checking, do we handle when suddenly someone disconnects/host disconnects/no one is left in the room, and we need to remove games/sessions?

When people leave while the game is running, the game state keeps running regardless and the game ends even when no people are in the lobby. In that case the  instance should be deleted.

However I just tested leaving the lobby while in the waiting room/before the game has started, and it seems the lobby is persisted regardless, will need to look at this further tomorrow.

**Review by jteo97 — APPROVED** [review]
Nice work.

**Comment by jteo97** [comment]
> > Looks great, thanks Leon :) Just double checking, do we handle when suddenly someone disconnects/host disconnects/no one is left in the room, and we need to remove games/sessions?
> 
> When people leave while the game is running, the game state keeps running regardless and the game ends even when no people are in the lobby. In that case the instance should be deleted.
> 
> However I just tested leaving the lobby while in the waiting room/before the game has started, and it seems the lobby is persisted regardless, will need to look at this further tomorrow.

Should be simple. Move the creation of the session from when the host first arrives onto the preference page to when they click the GO button.

**Comment by UltimaKuma** [comment]
> > > Looks great, thanks Leon :) Just double checking, do we handle when suddenly someone disconnects/host disconnects/no one is left in the room, and we need to remove games/sessions?
> > 
> > 
> > When people leave while the game is running, the game state keeps running regardless and the game ends even when no people are in the lobby. In that case the instance should be deleted.
> > However I just tested leaving the lobby while in the waiting room/before the game has started, and it seems the lobby is persisted regardless, will need to look at this further tomorrow.
> 
> Should be simple. Move the creation of the session from when the host first arrives onto the preference page to when they click the GO button.

Couldnt be done this way since the players joining the lobby need to reference the session, so it needs to exist when the lobby is created.

File: backend/src/sockets/leaveRoom.js [hunk 1]
```diff
@@ -13,7 +13,13 @@ export function leaveRoom(socket, io, cb = null) {
         const users = Array.from(
             game.session.users.values()).map((e) => e.name);
         console.log(users);
-        io.to(game.session.sessionId).emit('new_user', {users: users});
+        // Remove the game if the game hasnt started and no one is in the lobby
```
Comment (parfei) [inline backend/src/sockets/leaveRoom.js]: Nice Leon!

**Review by parfei — COMMENTED** [review]
image

SocketSession is not being removed even after going through a game, when you exit forcefully

**Comment by UltimaKuma** [comment]
Implementation was quite substantial and required the addition of an extra socket event as it was possible to exit the room in the preferences page. This meant that the room for the host needed to be deleted however there was no reference to which room belonged to which socket. Thus the socket event "host_room" was created to indicate who the host for the room was.

The back button for the preferences page was also modified to call the socket event "leave_room" in order for the room to be deleted

File: backend/src/sockets/disconnect.js [hunk 2]
```diff
@@ -1,23 +1,38 @@
...
+import games from "../domain/Games";
 /**
  * Socket event for when a user disconnects.
  * @param {*} socket
  * @param {*} io
  * @param {*} cb callback function
  */
 export function disconnect(socket, io, cb) {
-  socket.on('disconnect', () => {
+  socket.on("disconnect", () => {
```
Comment (parfei) [inline backend/src/sockets/disconnect.js]: a nitpick but eslint fail?

File: backend/src/sockets/leaveRoom.js [hunk 3]
```diff
@@ -1,24 +1,37 @@
...
  */
 export function leaveRoom(socket, io, cb = null) {
-  socket.on('leave_room', () => {
+  socket.on("leave_room", () => {
     const activeGames = games.getGames();
+
     activeGames.forEach((game) => {
+      let users = Array.from(game.session.users.values()).map((e) => e.name);
+      // Indicate host has left
+      if (game.session.hostSocket && socket.id === game.session.hostSocket.id) {
```
Comment (parfei) [inline backend/src/sockets/leaveRoom.js:15]: Nice!

File: backend/src/sockets/disconnect.js (hunk 2 above)
Comment (parfei) [reply backend/src/sockets/disconnect.js]: Might want to `npm run fix-lint`

File: backend/src/sockets/disconnect.js [hunk 4]
```diff
@@ -1,23 +1,38 @@
...
-            game.session.users.values()).map((e) => e.name);
+        // Recalculate users as have been changed
+        users = Array.from(game.session.users.values()).map((e) => e.name);
         console.log(users);
-        io.to(game.session.sessionId).emit('new_user', {users: users});
+        io.to(game.session.sessionId).emit("new_user", { users: users });
+      }
+      // Remove the game if the game hasnt started and no one is in the lobby
+      // To prevent a memory leak
+      console.log(users);
```
Comment (parfei) [inline backend/src/sockets/disconnect.js:28]: Could we remove this console log?

File: backend/src/sockets/disconnect.js [hunk 5]
```diff
@@ -1,23 +1,38 @@
...
+      }
+      // Remove users with same socket
       if (game.session.removeUser(socket)) {
-        const users = Array.from(
-            game.session.users.values()).map((e) => e.name);
+        // Recalculate users as have been changed
+        users = Array.from(game.session.users.values()).map((e) => e.name);
         console.log(users);
-        io.to(game.session.sessionId).emit('new_user', {users: users});
+        io.to(game.session.sessionId).emit("new_user", { users: users });
```
Comment (parfei) [inline backend/src/sockets/disconnect.js]: Another nitpick but not totally necessary: could extract this common functionality out

**Review by parfei — COMMENTED** [review]
Not sure if you care about this case (because it might be getting a bit too much, personally I'd say ignore it  and make a note somewhere :P), but if you leave during the game the SocketSession is not removed
image

Also, I think `SocketSession` is not removed when I complete the game too? Might be just me
image

Keen to hear your thoughts

Otherwise awesome work!! Might be worth adding to their Socket.IO API documentation in the wiki as well, since you add `host_room`

npm version: 7.6.3
platform: windows, chrome

**Comment by UltimaKuma** [comment]
> Not sure if you care about this case (because it might be getting a bit too much, personally I'd say ignore it and make a note somewhere :P), but if you leave during the game the SocketSession is not removed
> image
> 
> Also, I think `SocketSession` is not removed when I complete the game too? Might be just me
> image
> 
> Keen to hear your thoughts
> 
> Otherwise awesome work!! Might be worth adding to their Socket.IO API documentation in the wiki as well, since you add `host_room`
> 
> npm version: 7.6.3
> platform: windows, chrome

Wasnt able to recreate the error even with the same OS, npm version and node version. Will need to ignore for this PR at least.

As for leaving while the game is in progress, the session should be removed once the game has finished in the server.

**Review by parfei — APPROVED** [review]
Great work :) !! I feel like my Devtools is acting up, I console logged the `Games` object and cannot find the game object in there lol

**Review by jteo97 — APPROVED** [review]
Ran the game with these changes and it looks like it works. Both SocketSessions and Game are being deleted when there are no socket connections connected to that session. Tested with host leaving room but player still in it, etc.

**Pull request merged**